Add missing `[RequiresDynamicCode]` attributes to System.Linq.Expressions API in ref assembly

cston opened the pull request:
The ref assembly changes were generated by running `dotnet build /t:GenerateReferenceAssemblySource`, then reverting unnecessary differences and other API changes.
 
Fixes https://github.com/dotnet/runtime/issues/95077

cston committed:
dotnet build /t:GenerateReferenceAssemblySource

cston committed:
Revert unnecessary changes

cston committed:
Revert API changes

agocke reviewed (approved):
LGTM, thanks for the update

eerhardt commented:
I think you can remove these suppressions to fix the build:

https://github.com/dotnet/runtime/blob/2ed43b6a8105563f491fb521d761e6eff6ac4448/src/libraries/System.Linq.Expressions/src/CompatibilitySuppressions.xml#L100-L135

like the error says - "The APICompat suppression file can be updated by rebuilding with '/p:ApiCompatGenerateSuppressionFile=true'"

cston committed:
build ... /p:ApiCompatGenerateSuppressionFile=true

cston committed:
Update link

eerhardt commented on src/libraries/System.Linq.Expressions/src/CompatibilitySuppressions.xml:
```diff
@@ -1,5 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<!-- https://learn.microsoft.com/dotnet/fundamentals/package-validation/diagnostic-ids -->
+<!-- https://learn.microsoft.com/en-us/dotnet/fundamentals/package-validation/diagnostic-ids -->
```
```suggestion
<!-- https://learn.microsoft.com/dotnet/fundamentals/package-validation/diagnostic-ids -->
```

FYI - it looks like this was recently fixed with https://github.com/dotnet/sdk/pull/42830. We are just using an out of date version of the tool.

cc @ViktorHofer

eerhardt committed:
Resolve AOT warning in DependencyInjection ExpressionResolverBuilder

Calling Expression.Lambda without a delegate type isn't AOT compatible. Fix this by passing the delegate type.

eerhardt reviewed (approved):
Thanks for fixing this @cston!
